File: backend/core/src/modules/publishing-exports/publishing-exports.service.ts
import { BadRequestException, Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import crypto from 'crypto';
import JSZip from 'jszip';
import { ProjectAccessService } from '../../common/access/project-access.service';
import { ProjectScopedCrudService } from '../../common/access/project-scoped-crud.service';
import { Document, Project, PublishingExport } from '../../entities';

@Injectable()
export class PublishingExportsService extends ProjectScopedCrudService<PublishingExport> {
  constructor(
    @InjectRepository(PublishingExport)
    repo: Repository<PublishingExport>,
    @InjectRepository(Document)
    private readonly documentsRepo: Repository<Document>,
    @InjectRepository(Project)
    private readonly projectsRepo: Repository<Project>,
    projectAccessService: ProjectAccessService,
  ) {
    super(repo, projectAccessService);
  }

  private stripHtml(content: string): string {
    return content
      .replace(/<br\s*\/?>/gi, '\n')
      .replace(/<\/(p|div|h1|h2|h3|h4|h5|h6|li)>/gi, '\n')
      .replace(/<\/(tr)>/gi, '\n')
      .replace(/<\/(td|th)>/gi, ' ')
      .replace(/<[^>]*>/g, ' ')
      .replace(/&nbsp;/g, ' ')
      .replace(/\r\n/g, '\n')
      .replace(/[ \t]+\n/g, '\n')
      .replace(/\n{3,}/g, '\n\n')
      .replace(/[ \t]{2,}/g, ' ')
      .trim();
  }

  private safeFilename(name: string): string {
    return name
      .trim()
      .replace(/[\\/:*?"<>|]/g, '_')
      .replace(/\s+/g, '_')
      .slice(0, 80) || 'cowrite';
  }

  private async buildExport(exportJob: PublishingExport) {
    const project = await this.projectsRepo.findOne({
      where: { id: exportJob.projectId } as any,
      select: { title: true } as any,
    });
    const projectTitle = project?.title ?? 'cowrite';

    const docs = exportJob.documentId
      ? await this.documentsRepo.find({ where: { id: exportJob.documentId } as any })
      : await this.documentsRepo.find({
          where: { projectId: exportJob.projectId } as any,
          order: { orderIndex: 'ASC', createdAt: 'ASC' } as any,
        });

    if (docs.length === 0) {
      throw new BadRequestException('내보낼 문서가 없습니다.');
    }

    const sections = docs.map((d) => ({
      title: d.title,
      body: this.stripHtml(d.content ?? ''),
    }));

    const timestamp = new Date().toISOString().slice(0, 19).replace(/[-:T]/g, '');
    const base = `${this.safeFilename(projectTitle)}_${timestamp}`;

    switch (exportJob.exportFormat) {
      case 'txt': {
        const content = sections.map((s) => `${s.title}\n\n${s.body}`).join('\n\n---\n\n');
        return { content, filename: `${base}.txt`, mimeType: 'text/plain; charset=utf-8' };
      }
      case 'markdown': {
        const content = sections.map((s) => `# ${s.title}\n\n${s.body}`).join('\n\n---\n\n');
        return { content, filename: `${base}.md`, mimeType: 'text/markdown; charset=utf-8' };
      }
      case 'novelpia': {
        const content = sections.map((s) => `# ${s.title}\n\n${s.body}`).join('\n\n---\n\n');
        return { content, filename: `${base}.md`, mimeType: 'text/markdown; charset=utf-8' };
      }
      case 'kakaopage': {
        const content = sections
          .map((s) => `제목: ${s.title}\n\n${s.body}`)
          .join('\n\n\n');
        return { content: content.replace(/\n/g, '\r\n'), filename: `${base}.txt`, mimeType: 'text/plain; charset=utf-8' };
      }
      case 'munpia': {
        const content = sections
          .map((s) => `【${s.title}】\n\n${s.body}`)
          .join('\n\n\n');
        return { content: content.replace(/\n/g, '\r\n'), filename: `${base}.txt`, mimeType: 'text/plain; charset=utf-8' };
      }
      case 'epub': {
        const buffer = await this.buildEpub(projectTitle, sections);
        return {
          isBinary: true,
          contentBase64: buffer.toString('base64'),
          filename: `${base}.epub`,
          mimeType: 'application/epub+zip',
        };
      }
      default:
        throw new BadRequestException('지원하지 않는 exportFormat입니다. (txt, markdown, novelpia, kakaopage, munpia, epub)');
    }
  }

  private escapeXml(value: string): string {
    return value
      .replace(/&/g, '&amp;')
      .replace(/</g, '&lt;')
      .replace(/>/g, '&gt;')
      .replace(/"/g, '&quot;')
      .replace(/'/g, '&apos;');
  }

  private paragraphsToXhtml(body: string): string {
    if (!body) return '<p></p>';
    const parts = body.split(/\n{2,}/g).map((p) => p.trim()).filter(Boolean);
    const paragraphs = parts.map((p) => `<p>${this.escapeXml(p).replace(/\n/g, '<br />')}</p>`).join('\n');
    return paragraphs || '<p></p>';
  }

  private async buildEpub(title: string, sections: Array<{ title: string; body: string }>): Promise<Buffer> {
    const zip = new JSZip();

    zip.file('mimetype', 'application/epub+zip', { compression: 'STORE' });

    zip.folder('META-INF')?.file(
      'container.xml',
      `<?xml version="1.0" encoding="UTF-8"?>\n` +
        `<container version="1.0" xmlns="urn:oasis:names:tc:opendocument:xmlns:container">\n` +
        `  <rootfiles>\n` +
        `    <rootfile full-path="OEBPS/content.opf" media-type="application/oebps-package+xml"/>\n` +
        `  </rootfiles>\n` +
        `</container>\n`,
    );

    const oebps = zip.folder('OEBPS')!;
    const sectionsFolder = oebps.folder('sections')!;

    const sectionItems = sections.map((s, idx) => {
      const id = `sec${idx + 1}`;
      const href = `sections/${id}.xhtml`;
      const xhtml =
        `<?xml version="1.0" encoding="UTF-8"?>\n` +
        `<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="ko" lang="ko">\n` +
        `<head>\n` +
        `  <meta charset="utf-8" />\n` +
        `  <title>${this.escapeXml(s.title)}</title>\n` +
        `</head>\n` +
        `<body>\n` +
        `  <h1>${this.escapeXml(s.title)}</h1>\n` +
        `  ${this.paragraphsToXhtml(s.body)}\n` +
        `</body>\n` +
        `</html>\n`;
      sectionsFolder.file(`${id}.xhtml`, xhtml);
      return { id, href, title: s.title };
    });

    const navItems = sectionItems
      .map((item) => `<li><a href="${item.href}">${this.escapeXml(item.title)}</a></li>`)
      .join('\n');
    const navXhtml =
      `<?xml version="1.0" encoding="UTF-8"?>\n` +
      `<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="ko" lang="ko">\n` +
      `<head><meta charset="utf-8" /><title>목차</title></head>\n` +
      `<body>\n` +
      `  <nav epub:type="toc" xmlns:epub="http://www.idpf.org/2007/ops">\n` +
      `    <h1>목차</h1>\n` +
      `    <ol>\n${navItems}\n    </ol>\n` +
      `  </nav>\n` +
      `</body>\n` +
      `</html>\n`;
    oebps.file('nav.xhtml', navXhtml);

    const uuid = `urn:uuid:${crypto.randomUUID()}`;
    const manifestItems = [
      `<item id="nav" href="nav.xhtml" media-type="application/xhtml+xml" properties="nav"/>`,
      ...sectionItems.map((item) => `<item id="${item.id}" href="${item.href}" media-type="application/xhtml+xml"/>`),
    ].join('\n    ');

    const spineItems = sectionItems.map((item) => `<itemref idref="${item.id}"/>`).join('\n    ');

    const opf =
      `<?xml version="1.0" encoding="UTF-8"?>\n` +
      `<package xmlns="http://www.idpf.org/2007/opf" unique-identifier="bookid" version="3.0">\n` +
      `  <metadata xmlns:dc="http://purl.org/dc/elements/1.1/">\n` +
      `    <dc:identifier id="bookid">${uuid}</dc:identifier>\n` +
      `    <dc:title>${this.escapeXml(title)}</dc:title>\n` +
      `    <dc:language>ko</dc:language>\n` +
      `    <meta property="dcterms:modified">${new Date().toISOString()}</meta>\n` +
      `  </metadata>\n` +
      `  <manifest>\n    ${manifestItems}\n  </manifest>\n` +
      `  <spine>\n    ${spineItems}\n  </spine>\n` +
      `</package>\n`;
    oebps.file('content.opf', opf);

    return zip.generateAsync({ type: 'nodebuffer', compression: 'DEFLATE' });
  }

  override async createForUser(userId: string | undefined, dto: Partial<PublishingExport>): Promise<PublishingExport> {
    const created = await super.createForUser(userId, { ...dto, status: 'processing' });

    try {
      const built = await this.buildExport(created);
      const fileUrl = `/publishing-exports/${created.id}/download`;
      const metadata = {
        ...(created.metadata ?? {}),
        ...built,
      } as Record<string, unknown>;

      await this.repo.update(
        { id: created.id } as any,
        {
          status: 'completed',
          completedAt: new Date(),
          fileUrl,
          metadata,
        } as any,
      );

      const updated = await this.findOneForUser(userId, created.id);
      if (!updated) {
        throw new BadRequestException('내보내기 결과를 찾을 수 없습니다.');
      }
      return updated;
    } catch (error) {
      const message = error instanceof Error ? error.message : '내보내기에 실패했습니다.';
      await this.repo.update(
        { id: created.id } as any,
        {
          status: 'failed',
          metadata: { ...(created.metadata ?? {}), error: message } as any,
        } as any,
      );
      throw error;
    }
  }

  async getDownloadForUser(userId: string | undefined, id: string) {
    const exportJob = await this.findOneForUser(userId, id);
    if (!exportJob) {
      throw new BadRequestException('내보내기 항목을 찾을 수 없습니다.');
    }
    if (exportJob.status !== 'completed') {
      throw new BadRequestException('내보내기가 완료되지 않았습니다.');
    }

    const metadata = exportJob.metadata ?? {};
    const isBinary = Boolean((metadata as any).isBinary);
    let content: string | Buffer = '';
    if (isBinary) {
      const b64 = typeof (metadata as any).contentBase64 === 'string' ? ((metadata as any).contentBase64 as string) : '';
      if (!b64) throw new BadRequestException('내보내기 파일이 없습니다.');
      content = Buffer.from(b64, 'base64');
    } else {
      content = typeof (metadata as any).content === 'string' ? ((metadata as any).content as string) : '';
      if (!content) throw new BadRequestException('내보내기 파일이 없습니다.');
    }

    const filename =
      typeof (metadata as any).filename === 'string' ? ((metadata as any).filename as string) : `cowrite-export.${exportJob.exportFormat}`;
    const mimeType =
      typeof (metadata as any).mimeType === 'string' ? ((metadata as any).mimeType as string) : 'application/octet-stream';

    return { filename, mimeType, content };
  }
}

File: backend/core/src/main/java/store/pjcloud/cowrite/core/service/PublishingExportsService.java
package store.pjcloud.cowrite.core.service;

import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.time.OffsetDateTime;
import java.util.Base64;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.zip.CRC32;
import java.util.zip.ZipEntry;
import java.util.zip.ZipOutputStream;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;
import store.pjcloud.cowrite.core.common.ProjectScopedCrudService;
import store.pjcloud.cowrite.core.entity.Document;
import store.pjcloud.cowrite.core.entity.Project;
import store.pjcloud.cowrite.core.entity.PublishingExport;
import store.pjcloud.cowrite.core.repository.DocumentRepository;
import store.pjcloud.cowrite.core.repository.ProjectRepository;
import store.pjcloud.cowrite.core.repository.PublishingExportRepository;

@Service
public class PublishingExportsService extends ProjectScopedCrudService<PublishingExport> {
    private final DocumentRepository documentsRepo;
    private final ProjectRepository projectsRepo;
    private final ObjectMapper objectMapper;

    public PublishingExportsService(PublishingExportRepository repository,
                                    ProjectAccessService projectAccessService,
                                    ObjectMapper objectMapper,
                                    DocumentRepository documentsRepo,
                                    ProjectRepository projectsRepo) {
        super(repository, projectAccessService, objectMapper, PublishingExport.class);
        this.documentsRepo = documentsRepo;
        this.projectsRepo = projectsRepo;
        this.objectMapper = objectMapper;
    }

    @Override
    @Transactional
    public PublishingExport createForUser(UUID userId, Map<String, Object> body) {
        PublishingExport created = super.createForUser(userId, body);
        created.setStatus("processing");
        repository().save(created);

        try {
            ExportResult built = buildExport(created);
            String fileUrl = "/publishing-exports/" + created.getId() + "/download";
            Map<String, Object> metadata = objectMapper.convertValue(created.getMetadata(), Map.class);
            if (metadata == null) metadata = new java.util.HashMap<>();
            metadata.putAll(built.metadata());

            created.setStatus("completed");
            created.setCompletedAt(OffsetDateTime.now());
            created.setFileUrl(fileUrl);
            created.setMetadata(metadata);
            repository().save(created);
            return created;
        } catch (Exception ex) {
            Map<String, Object> metadata = objectMapper.convertValue(created.getMetadata(), Map.class);
            if (metadata == null) metadata = new java.util.HashMap<>();
            metadata.put("error", ex.getMessage() == null ? "내보내기에 실패했습니다." : ex.getMessage());
            created.setStatus("failed");
            created.setMetadata(metadata);
            repository().save(created);
            throw ex instanceof RuntimeException ? (RuntimeException) ex : new RuntimeException(ex);
        }
    }

    public DownloadResult getDownloadForUser(UUID userId, UUID id) {
        PublishingExport exportJob = findOneForUser(userId, id);
        if (exportJob == null) throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "내보내기 항목을 찾을 수 없습니다.");
        if (!"completed".equals(exportJob.getStatus())) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "내보내기가 완료되지 않았습니다.");
        }
        Map<String, Object> metadata = exportJob.getMetadata();
        boolean isBinary = metadata != null && Boolean.TRUE.equals(metadata.get("isBinary"));
        byte[] content;
        if (isBinary) {
            String b64 = metadata != null && metadata.get("contentBase64") instanceof String s ? s : "";
            if (b64.isBlank()) throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "내보내기 파일이 없습니다.");
            content = Base64.getDecoder().decode(b64);
        } else {
            String text = metadata != null && metadata.get("content") instanceof String s ? s : "";
            if (text.isBlank()) throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "내보내기 파일이 없습니다.");
            content = text.getBytes(StandardCharsets.UTF_8);
        }

        String filename = metadata != null && metadata.get("filename") instanceof String s ? s : ("cowrite-export." + exportJob.getExportFormat());
        String mimeType = metadata != null && metadata.get("mimeType") instanceof String s ? s : "application/octet-stream";
        return new DownloadResult(filename, mimeType, content);
    }

    private ExportResult buildExport(PublishingExport exportJob) {
        Project project = projectsRepo.findById(exportJob.getProjectId()).orElse(null);
        String projectTitle = project != null ? project.getTitle() : "cowrite";

        List<Document> docs = exportJob.getDocumentId() != null
            ? documentsRepo.findAllById(List.of(exportJob.getDocumentId()))
            : documentsRepo.findByProjectIdOrderByOrderIndexAscCreatedAtAsc(exportJob.getProjectId());

        if (docs.isEmpty()) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "내보낼 문서가 없습니다.");
        }

        List<Section> sections = docs.stream().map(d -> new Section(d.getTitle(), stripHtml(d.getContent() == null ? "" : d.getContent()))).toList();
        String timestamp = java.time.OffsetDateTime.now().toString().replaceAll("[-:T]", "").substring(0, 14);
        String base = safeFilename(projectTitle) + "_" + timestamp;

        return switch (exportJob.getExportFormat()) {
            case "txt" -> new ExportResult(Map.of(
                "content", sectionsToText(sections, "\n\n---\n\n"),
                "filename", base + ".txt",
                "mimeType", "text/plain; charset=utf-8"
            ));
            case "markdown", "novelpia" -> new ExportResult(Map.of(
                "content", sectionsToMarkdown(sections),
                "filename", base + ".md",
                "mimeType", "text/markdown; charset=utf-8"
            ));
            case "kakaopage" -> new ExportResult(Map.of(
                "content", sectionsToKakao(sections),
                "filename", base + ".txt",
                "mimeType", "text/plain; charset=utf-8"
            ));
            case "munpia" -> new ExportResult(Map.of(
                "content", sectionsToMunpia(sections),
                "filename", base + ".txt",
                "mimeType", "text/plain; charset=utf-8"
            ));
            case "epub" -> {
                byte[] buffer = buildEpub(projectTitle, sections);
                yield new ExportResult(Map.of(
                    "isBinary", true,
                    "contentBase64", Base64.getEncoder().encodeToString(buffer),
                    "filename", base + ".epub",
                    "mimeType", "application/epub+zip"
                ));
            }
            default -> throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "지원하지 않는 exportFormat입니다. (txt, markdown, novelpia, kakaopage, munpia, epub)");
        };
    }

    private String stripHtml(String content) {
        return content
            .replaceAll("<br\\s*/?>", "\n")
            .replaceAll("</(p|div|h1|h2|h3|h4|h5|h6|li)>", "\n")
            .replaceAll("</(tr)>", "\n")
            .replaceAll("</(td|th)>", " ")
            .replaceAll("<[^>]*>", " ")
            .replace("&nbsp;", " ")
            .replaceAll("\\r\\n", "\n")
            .replaceAll("[ \t]+\n", "\n")
            .replaceAll("\n{3,}", "\n\n")
            .replaceAll("[ \t]{2,}", " ")
            .trim();
    }

    private String safeFilename(String name) {
        String cleaned = name == null ? "" : name.trim().replaceAll("[\\\\/:*?\"<>|]", "_").replaceAll("\\s+", "_");
        if (cleaned.isBlank()) return "cowrite";
        return cleaned.length() > 80 ? cleaned.substring(0, 80) : cleaned;
    }

    private String sectionsToText(List<Section> sections, String separator) {
        return sections.stream().map(s -> s.title() + "\n\n" + s.body()).collect(java.util.stream.Collectors.joining(separator));
    }

    private String sectionsToMarkdown(List<Section> sections) {
        return sections.stream().map(s -> "# " + s.title() + "\n\n" + s.body()).collect(java.util.stream.Collectors.joining("\n\n---\n\n"));
    }

    private String sectionsToKakao(List<Section> sections) {
        String content = sections.stream().map(s -> "제목: " + s.title() + "\n\n" + s.body()).collect(java.util.stream.Collectors.joining("\n\n\n"));
        return content.replace("\n", "\r\n");
    }

    private String sectionsToMunpia(List<Section> sections) {
        String content = sections.stream().map(s -> "【" + s.title() + "】\n\n" + s.body()).collect(java.util.stream.Collectors.joining("\n\n\n"));
        return content.replace("\n", "\r\n");
    }

    private byte[] buildEpub(String title, List<Section> sections) {
        try {
            ByteArrayOutputStream baos = new ByteArrayOutputStream();
            ZipOutputStream zos = new ZipOutputStream(baos);

            writeStoredEntry(zos, "mimetype", "application/epub+zip");
            writeEntry(zos, "META-INF/container.xml", (
                "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n" +
                "<container version=\"1.0\" xmlns=\"urn:oasis:names:tc:opendocument:xmlns:container\">\n" +
                "  <rootfiles>\n" +
                "    <rootfile full-path=\"OEBPS/content.opf\" media-type=\"application/oebps-package+xml\"/>\n" +
                "  </rootfiles>\n" +
                "</container>\n"
            ));

            StringBuilder manifestItems = new StringBuilder();
            StringBuilder spineItems = new StringBuilder();
            StringBuilder navItems = new StringBuilder();

            int idx = 1;
            for (Section section : sections) {
                String id = "sec" + idx;
                String href = "sections/" + id + ".xhtml";
                String xhtml = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n" +
                    "<html xmlns=\"http://www.w3.org/1999/xhtml\" xml:lang=\"ko\" lang=\"ko\">\n" +
                    "<head><meta charset=\"utf-8\"/><title>" + escapeXml(section.title()) + "</title></head>\n" +
                    "<body>\n  <h1>" + escapeXml(section.title()) + "</h1>\n  " + paragraphsToXhtml(section.body()) + "\n</body>\n</html>\n";
                writeEntry(zos, "OEBPS/" + href, xhtml);
                manifestItems.append("<item id=\"").append(id).append("\" href=\"").append(href).append("\" media-type=\"application/xhtml+xml\"/>\n    ");
                spineItems.append("<itemref idref=\"").append(id).append("\"/>\n    ");
                navItems.append("<li><a href=\"").append(href).append("\">").append(escapeXml(section.title())).append("</a></li>\n");
                idx++;
            }

            String navXhtml = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n" +
                "<html xmlns=\"http://www.w3.org/1999/xhtml\" xml:lang=\"ko\" lang=\"ko\">\n" +
                "<head><meta charset=\"utf-8\" /><title>목차</title></head>\n" +
                "<body>\n  <nav epub:type=\"toc\" xmlns:epub=\"http://www.idpf.org/2007/ops\">\n    <h1>목차</h1>\n    <ol>\n" +
                navItems + "    </ol>\n  </nav>\n</body>\n</html>\n";
            writeEntry(zos, "OEBPS/nav.xhtml", navXhtml);

            String uuid = "urn:uuid:" + java.util.UUID.randomUUID();
            String opf = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n" +
                "<package xmlns=\"http://www.idpf.org/2007/opf\" unique-identifier=\"bookid\" version=\"3.0\">\n" +
                "  <metadata xmlns:dc=\"http://purl.org/dc/elements/1.1/\">\n" +
                "    <dc:identifier id=\"bookid\">" + uuid + "</dc:identifier>\n" +
                "    <dc:title>" + escapeXml(title) + "</dc:title>\n" +
                "    <dc:language>ko</dc:language>\n" +
                "    <meta property=\"dcterms:modified\">" + java.time.OffsetDateTime.now().toString() + "</meta>\n" +
                "  </metadata>\n" +
                "  <manifest>\n    <item id=\"nav\" href=\"nav.xhtml\" media-type=\"application/xhtml+xml\" properties=\"nav\"/>\n    " + manifestItems + "  </manifest>\n" +
                "  <spine>\n    " + spineItems + "  </spine>\n" +
                "</package>\n";
            writeEntry(zos, "OEBPS/content.opf", opf);

            zos.close();
            return baos.toByteArray();
        } catch (IOException ex) {
            throw new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR, "EPUB 생성에 실패했습니다.");
        }
    }

    private void writeEntry(ZipOutputStream zos, String path, String content) throws IOException {
        ZipEntry entry = new ZipEntry(path);
        zos.putNextEntry(entry);
        zos.write(content.getBytes(StandardCharsets.UTF_8));
        zos.closeEntry();
    }

    private void writeStoredEntry(ZipOutputStream zos, String path, String content) throws IOException {
        byte[] bytes = content.getBytes(StandardCharsets.UTF_8);
        CRC32 crc = new CRC32();
        crc.update(bytes);
        ZipEntry entry = new ZipEntry(path);
        entry.setMethod(ZipEntry.STORED);
        entry.setSize(bytes.length);
        entry.setCompressedSize(bytes.length);
        entry.setCrc(crc.getValue());
        zos.putNextEntry(entry);
        zos.write(bytes);
        zos.closeEntry();
    }

    private String escapeXml(String value) {
        if (value == null) return "";
        return value.replace("&", "&amp;")
            .replace("<", "&lt;")
            .replace(">", "&gt;")
            .replace("\"", "&quot;")
            .replace("'", "&apos;");
    }

    private String paragraphsToXhtml(String body) {
        if (body == null || body.isBlank()) return "<p></p>";
        String[] parts = body.split("\\n{2,}");
        StringBuilder sb = new StringBuilder();
        for (String part : parts) {
            String p = part.trim();
            if (p.isEmpty()) continue;
            sb.append("<p>").append(escapeXml(p).replace("\n", "<br />")).append("</p>\n");
        }
        return sb.length() == 0 ? "<p></p>" : sb.toString();
    }

    private record Section(String title, String body) {}
    private record ExportResult(Map<String, Object> metadata) {}
    public record DownloadResult(String filename, String mimeType, byte[] content) {}
}
